test/chi0.jl

```diff
@@ -4,92 +4,103 @@ using LinearAlgebra: norm
+        occ, εF = DFTK.compute_occupation(basis, res.eigenvalues)
+        scfres = (ham=ham0, res..., n_ep_extra=n_ep_extra)
 
-        # Now we make the same model, but add an artificial external potential ε * δV
+        # create external small perturbation εδV
         n_spin = model.n_spin_components
         δV = randn(eltype(basis), basis.fft_size..., n_spin)
-        δV_sym = DFTK.symmetrize_ρ(basis, δV, symmetries=model.symmetries)
+        δV_sym = DFTK.symmetrize_ρ(basis, δV)
         if symmetry
             δV = δV_sym
         else
             @test δV_sym ≈ δV
         end
-
         εδV = ε * δV
+
+        # middle finite difference for more precision
+
+        term_builder = basis -> DFTK.TermExternal(-εδV)
+        model = model_LDA(testcase.lattice, [spec => testcase.positions];
+                          model_kwargs..., extra_terms=[term_builder])
+        basis = PlaneWaveBasis(model; Ecut, basis_kwargs...)
+        energies, ham = energy_hamiltonian(basis, nothing, nothing; ρ=ρ0)
+        res = DFTK.next_density(ham, tol=tol, eigensolver=lobpcg_hyper,
+                                n_ep_extra=n_ep_extra)
+        ρ1 = res.ρout
+
         term_builder = basis -> DFTK.TermExternal(εδV)
         model = model_LDA(testcase.lattice, [spec => testcase.positions];
                           model_kwargs..., extra_terms=[term_builder])
         basis = PlaneWaveBasis(model; Ecut, basis_kwargs...)
         energies, ham = energy_hamiltonian(basis, nothing, nothing; ρ=ρ0)
-        res = DFTK.next_density(ham, tol=tol, eigensolver=diag_full, n_bands=n_bands)
-        ρ2     = res.ρout
-        diff_findiff = (ρ2 - ρ1) / ε
+        res = DFTK.next_density(ham, tol=tol, eigensolver=lobpcg_hyper,
+                                n_ep_extra=n_ep_extra)
+        ρ2 = res.ρout
 
-        EVs = [eigen(Hermitian(Array(Hk))) for Hk in ham0.blocks]
-        Es = [EV.values[1:n_bands] for EV in EVs]
-        Vs = [EV.vectors[:, 1:n_bands] for EV in EVs]
-        occ, εF = DFTK.compute_occupation(basis, Es)
+        diff_findiff = (ρ2 - ρ1) / (2ε)
 
         # Test apply_χ0 and compare against finite differences
-        diff_applied_χ0 = apply_χ0(ham0, Vs, εF, Es, δV)
+        diff_applied_χ0 = apply_χ0(scfres, δV)
         @test norm(diff_findiff - diff_applied_χ0) < testtol
 
         # just to cover it here
         if temperature > 0
-            N = compute_nos(εF, basis, Es)
-            D = compute_dos(εF, basis, Es)
-            LDOS = compute_ldos(εF, basis, Es, Vs)
+            N = compute_nos(εF, basis, res.eigenvalues)
+            D = compute_dos(εF, basis, res.eigenvalues)
+            LDOS = compute_ldos(εF, basis, res.eigenvalues, res.ψ)
         end
 
         if !symmetry
-            # Test compute_χ0 against finite differences
-            χ0 = compute_χ0(ham0)
-            diff_computed_χ0 = reshape(χ0 * vec(δV), basis.fft_size..., n_spin)
-            @test norm(diff_findiff - diff_computed_χ0) < testtol
-
```

Comment:
Yep, I agree. But then we have to find a compromise between small Ecut and small fft_size so that the compuation is not too long and high Ecut so that LOBPCG doesn't crash. 

For the moment, I included this test for small Ecut and fft_size, but replace the LOBPCG by full diagonlization, but then the extra eigenvectors are fully converged.

Reply:
You can just make a flag to run the `compute_χ0` only in some of the cases, for which you lower `Ecut` and `fft_size` explictly and switch the solver to full (just do this via kwargs to the test function).

The function is pretty simple and not as important as the `apply_χ0`, so you don't need to test it so exhaustively wrt. metals/insulators etc. Just cover a reasonable subset.